libyul/AsmAnalysis.cpp

```diff
@@ -772,10 +772,27 @@ bool AsmAnalyzer::validateInstructions(evmasm::Instruction _instr, SourceLocatio
 			"PC instruction is a low-level EVM feature. "
 			"Because of that PC is disallowed in strict assembly."
 		);
+	else if (!m_eofVersion.has_value() && (
+		_instr == evmasm::Instruction::EXTCALL ||
+		_instr == evmasm::Instruction::EXTDELEGATECALL ||
+		_instr == evmasm::Instruction::EXTSTATICCALL
+	))
+	{
+		m_errorReporter.typeError(
+			4328_error,
+			_location,
+			fmt::format(
+				"The \"{instruction}\" instruction is {kind} VMs.",
+				fmt::arg("instruction", boost::to_lower_copy(instructionInfo(_instr, m_evmVersion).name)),
+				fmt::arg("kind", "not available in legacy bytecode")
```

Comment:
Because these three opcodes are available in yul (EOF) as regular instructions. Other instructions I added are buildin functions or are unavailable in yul. If we don't add [the sanity check](https://github.com/ethereum/solidity/blob/develop/libyul/AsmAnalysis.cpp#L804) fails

Reply:
I mean all added opcodes were till now buildin or are disabled in dialect.

Reply:
Ah, right. I'm still confusing this bit with the code responsible for making sure instructions are available or not on EOF/legacy. But this only kicks in when we would get an error anyway and just replaces the generic "function not found" with a more specific message explaining why it was not available.

So this is correct, it's the other way around - the dialect builtins should report these specialized messages to if we wanted to be fully consistent.

Reply:
But while thinking about it I realized we're missing a check elsewhere and a very important one - we're missing the exceptions for the `EXT*CALL` instructions on legacy. They should be marked as non-reserved in `EVMDialect::createReservedIdentifiers()`. Without it, this is breaking for legacy Yul.

Reply:
And it's actually not just this PR, we're missing such exceptions for nearly all the instructions we added so far. We did that only for `auxdataloadn`, but then we forgot `dataloadn`, which gets added automatically as all `Instruction`s are.

This thing is just too error-prone and we keep realizing we're missing something there (https://github.com/ethereum/solidity/pull/15550#discussion_r1859401044). I think that instead of testing it ad-hoc for every new opcode/builtin, we should create a few tests that cover all cases for all of them and just keep them updated.

We basically need 4 tests. Each one should cover all EOF-only opcodes and all EOF-only builtins and check that:
1. The name is **not** reserved on legacy (can declare variables and functions with that name). Covers `createReservedIdentifiers()`.
2. The name is reserved on EOF.
3. The name is **not** defined on legacy (calling that name will not execute a Yul builtin). Covers `createBuiltins()`.
    - If it works on EOF, we get a specialized error saying it's only available on EOF (rather than a generic error about calling undefined function). Covers `validateInstructions()` we have here.
4. It either works on EOF or is not defined (depending on the case).

Reply:
Addressed. I created these test covering `ext*call` in this PR. They should be extended to all added EOF instructions